op_builder/builder.py

```diff
@@ -0,0 +1,206 @@
+            )
+        if not self.command_exists('ninja'):
+            raise RuntimeError(
+                f"Unable to JIT load the {self.name} op due to ninja not being installed."
+            )
+
+        self.jit_mode = True
+        from torch.utils.cpp_extension import load
+
+        # Ensure directory exists to prevent race condition in some cases
+        ext_path = os.path.join(
+            os.environ.get('TORCH_EXTENSIONS_DIR',
+                           DEFAULT_TORCH_EXTENSION_PATH),
+            self.name)
+        os.makedirs(ext_path, exist_ok=True)
+
+        start_build = time.time()
+        op_module = load(
+            name=self.name,
+            sources=[self.deepspeed_src_path(path) for path in self.sources()],
+            extra_include_paths=[
+                self.deepspeed_src_path(path) for path in self.include_paths()
+            ],
+            extra_cflags=self.cxx_args(),
+            extra_cuda_cflags=self.nvcc_args(),
+            verbose=verbose)
+        build_duration = time.time() - start_build
+        if verbose:
+            print(f"Time to load {self.name} op: {build_duration} seconds")
+        return op_module
+
+
+class CUDAOpBuilder(OpBuilder):
+    def compute_capability_args(self, cross_compile_archs=['52', '60', '61', '70']):
+        args = []
+        if self.jit_mode:
+            # Compile for underlying architecture since we know it at runtime
+            CC_MAJOR, CC_MINOR = torch.cuda.get_device_capability()
+            compute_capability = f"{CC_MAJOR}{CC_MINOR}"
+            args.append('-gencode')
+            args.append(
+                f'arch=compute_{compute_capability},code=compute_{compute_capability}')
+        else:
+            # Cross-compile mode, compile for various architectures
+            for compute_capability in cross_compile_archs:
+                args.append('-gencode')
+                args.append(
+                    f'arch=compute_{compute_capability},code=compute_{compute_capability}'
+                )
+        return args
+
+    def version_dependent_macros(self):
+        # Fix from apex that might be relevant for us as well, related to https://github.com/NVIDIA/apex/issues/456
+        TORCH_MAJOR = int(torch.__version__.split('.')[0])
+        TORCH_MINOR = int(torch.__version__.split('.')[1])
+        version_ge_1_1 = []
+        if (TORCH_MAJOR > 1) or (TORCH_MAJOR == 1 and TORCH_MINOR > 0):
+            version_ge_1_1 = ['-DVERSION_GE_1_1']
+        version_ge_1_3 = []
+        if (TORCH_MAJOR > 1) or (TORCH_MAJOR == 1 and TORCH_MINOR > 2):
+            version_ge_1_3 = ['-DVERSION_GE_1_3']
+        version_ge_1_5 = []
+        if (TORCH_MAJOR > 1) or (TORCH_MAJOR == 1 and TORCH_MINOR > 4):
+            version_ge_1_5 = ['-DVERSION_GE_1_5']
+        return version_ge_1_1 + version_ge_1_3 + version_ge_1_5
+
+    def is_compatible(self):
+        assert_no_cuda_mismatch()
```

Comment:
I totally agree that a version mismatch should be a showstopper when we go to build an op. However, asserting here in the compatibility check seems very aggressive and stops us installing on non-CUDA systems.